File: Blog/Algorithm/Top20CodingInterview/Q.01.Majority/Q.01.Majority.py
class NoMajority(Exception):
    def __init__(self):
        self.value = "No Majority"

    def __str__(self):
        return self.value

def FindMajority(listNumbers):
    mExsitingNumbers = dict()
    nMajority = len(listNumbers) / 2
    for nNumber in listNumbers:
        if nNumber in mExsitingNumbers:
            mExsitingNumbers[nNumber] += 1
            if mExsitingNumbers[nNumber] >= nMajority:
                return nNumber
        else:
            mExsitingNumbers[nNumber] = 1
    raise NoMajority

listNumbers = [ 1,2,3,4,5,6,7,8,9,0]

try:
    print(FindMajority(listNumbers))
except NoMajority as err:
    print(err)

try:
    print(FindMajority([1,2,3,4,4,4,4,6,3,4]))
except NoMajority as err:
    print(err)

File: Blog/Algorithm/Top20CodingInterview/Q.01.Majority/Q.01.Majority.java
import java.util.*;

public class Q01_Majority {

	public static int FindMajority(List<Integer> vNumbers) throws Exception
	{
		int nSize = vNumbers.size();
		int nMajority = nSize / 2;
		HashMap<Integer, Integer> mExistingNumbers = new HashMap<Integer, Integer>();
		
		for (int i = 0; i < nSize; i++)
		{
			int nNumber = vNumbers.get(i);
			Integer nValue = mExistingNumbers.get(nNumber);
			if (nValue != null)
			{
				nValue++;
				if (nValue >= nMajority)
					return nNumber;
				mExistingNumbers.put(nNumber, nValue);
			}
			else
			{
				mExistingNumbers.put(nNumber, 1);
			}
		}

		throw new Exception("No Majority");
	}
	
	public static void main(String[] args) 
	{
		List<Integer> nNumbers1 = Arrays.asList(1,2,3,4,4,4,4,4,3,3);
		List<Integer> nNumbers2 = Arrays.asList(1,2,3,4,5,6,7,8,9,0);
		
		try 
		{
			System.out.println(FindMajority(nNumbers1));
			System.out.println(FindMajority(nNumbers2));
		}
		catch (Exception e)
		{
			System.out.println(e.getMessage());
		}
	}
}
